### ISSUE TITLE
Distributed Training Is Not Working

### ISSUE THREAD

**Issue by alexlimh** [facebookresearch/DPR#80]
Hi,

I tried training the dense encoder on 2 V100 GPUs using the following commands:
`python -m torch.distributed.launch \
	--nproc_per_node=2 train_dense_encoder.py \
	--max_grad_norm 2.0 \
	--encoder_model_type hf_bert \
	--pretrained_model_cfg bert-base-uncased \
	--seed 12345 \
	--sequence_length 256 \
	--warmup_steps 1237 \
	--batch_size 16 \
	--do_lower_case \
	--train_file "data/retriever/nq-train.json" \
	--dev_file "data/retriever/nq-dev.json" \
	--output_dir logdir \
	--learning_rate 2e-05 \
	--num_train_epochs 40 \
	--dev_batch_size 16 \
	--val_av_rank_start_epoch 30`
 but the model doesn't seem to be trained at all during the 40 epochs:
Epoch 0 and 1:
![image](https://user-images.githubusercontent.com/19368423/98727170-84b42680-2365-11eb-94d3-4082475c732f.png)
Epoch 28 and 29:
![image](https://user-images.githubusercontent.com/19368423/98727637-3a7f7500-2366-11eb-82e0-97a1977d9276.png)

I would appreciate it if you could tell me where's the problem cuz I think I might miss something...

Many thanks,
Minghan

**Comment by vlad-karpukhin** [comment]
Hi Minghan,
You use just two gpus with the batch size 16, so your overall effective batch size is 2x16=32
Biencoder model with "in-bacth negatives" scheme's efficiency depends directly on the effective batch size. 32 is quite low, we used 128 by using 8 gpus. 
Unfortunately, you won't be able to train the retriever model to our quality using just 2 gpus.

However, It is strange that your NLL validation loss is not improving at all.
Can you check if the train set loss per epoch  and rolling train loss values are  decreasing in the logs?

**Comment by alexlimh** [comment]
Hi Vladimir,

Thanks for your quick response. I understand that using a bigger batch size might improve the performance, but I just don't have enough compute to do that.

I checked the training loss per epoch and it's always something like 
"Av Loss per epoch=0.000000
epoch total correct predictions=0" 

However, when I trained the model on only 1 GPU, both the training loss and validation loss seemed fine to me, as well as the final results. Any clues?

**Comment by vlad-karpukhin** [comment]
So your train loss is 0 and no correct predictions reported. There is something wring going on.
Can you upload entire log file somewhere so I can check it for more details?

**Comment by alexlimh** [comment]
Here:
[train.log](https://github.com/facebookresearch/DPR/files/5520253/train.log)
Thanks

**Comment by vlad-karpukhin** [comment]
I don't see that you model read the train data file and actually do any training
"""
  Total iterations per epoch=0
 Total updates=0
  Eval step = 0
"""
There is no training at all.

Not sure what went wrong but looks like your python code can't find your train file.
Can you check that it actually exists and try to provide the full path to the file (vs the relative one)?

**Comment by alexlimh** [comment]
Ah, you are right. I checked my `data` directory and it turns out there's only an `nq-train.tmp` file instead of the `nq-train.json` file, which might be the reason for the "not training" problem.

However, I find it weird that there was no error or warning during downloading the data or training. I think it would be better to stop the program if there are no training data detected, instead of letting it keep running.

**Comment by vlad-karpukhin** [comment]
valid point. 
The current code assumes you want just to evaluate your model when no train data are provided and just call 2 validate methods. But it doesn't handle the case when you actually provide train data but it points to a non existent file(s).

**Comment by vlad-karpukhin** [comment]
Fixed in the new release
